oxide-auth/src/primitives/generator.rs

```diff
@@ -44,180 +36,68 @@ pub trait TagGrant {
     fn tag(&mut self, usage: u64, grant: &Grant) -> Result<String, ()>;
 }
 
-/// Generates tokens from random bytes.
-///
-/// Each byte is chosen randomly from the basic `rand::thread_rng`. This generator will always
-/// succeed.
-pub struct RandomGenerator {
-    random: SystemRandom,
-    len: usize
+/// An implementation of TagGrant backed by a cryptographically secure source of randomness
+pub struct RandGenerator<T> {
```

Comment:
Having `rand` implementations makes the basic interfaces more useful. The names don't need to be differ and I prefer having a single, known good implementation to overly configurable systems. What is gained from the parameter? The instantiation should be a perfectly unpredictable generator by design, all of which should be indistinguishable if well seeded. This only adds a point of failure. To quote from [`rand::CryptoRng`](https://docs.rs/rand/0.7.2/rand/trait.CryptoRng.html) which acknowledges this issue:

> Note that this trait is provided for guidance only and cannot guarantee suitability for cryptographic applications. In general it should only be implemented for well-reviewed code implementing well-regarded algorithms.

This seems to suggest that it is *unsafe* to allow generic instantiations.

Reply:
Note: A renaming to `WeakRandGenerator` may be fine. At least it makes the potential consequences and the level of scrutiny applied to the implementation clear from the start.

Reply:
What if instead of making it generic over CryptoRng I put a Box<dyn Rng> in it and only exposed methods for creating with StdRng (Send+Sync) or ThreadRng (!Send+!Sync)

Reply:
Maybe with a method like `fn unsafe_custom_rng<R: CryptoRng + Rng>(&mut self, rng: R)`

Reply:
I guess boxing wouldn't work with the sync amd send bounds, though

Reply:
Just going with StdRng might be the best option, since it's Send+Sync, part of the Rand crate, and officially marked CryptoRng

Reply:
That were my thoughts as well if there is no obvious advantage of `ThreadRng`.